### PR TITLE
Update to librustzcash version 0.7

### PR DESCRIPTION
**Pull request by nuttycom**

### CODE REVIEW

**Comment by pacu** [comment]
I think we can unblock this by unscoping [Remove accountId from UnifiedFullViewingKey FFI representation](https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/41)

@nuttycom @str4d

**Comment by pacu** [comment]
@str4d @nuttycom  can we remove [Remove accountId from UnifiedFullViewingKey FFI representation](https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/41) from this PR's blockers?

File: rust/Cargo.toml
Comment (str4d) [inline rust/Cargo.toml:46]: `orchard` also needs to be patched:
```suggestion
schemer-rusqlite = { git = "https://github.com/aschampion/schemer.git", rev = "6726b60f43f72c6e24a18d31be0ec7d42829e5e1" }
orchard = { git = "https://github.com/zcash/orchard.git", rev = "4035a97d178f99bb889f4d7ed26c59378fa88961" }
```

File: CHANGELOG.md
Comment (pacu) [inline CHANGELOG.md:37]: @nuttycom  we are missing the get metadata entry on the changelog

**Review by pacu — CHANGES_REQUESTED** [review]
@nuttycom we are missing the get metadata entry on the changelog

File: rust/src/lib.rs
Comment (str4d) [inline rust/src/lib.rs:2149]: Given that this is named `opt_anchor`, this extracts the wrong value; the tuple is `(target, anchor)`. Then again, with a confirmations of `0`, the two should be identical.

Comment (sellout) [reply rust/src/lib.rs:2149]: There are two other occurrences of `opt_anchor` that then extract the target height instead of the anchor height. Are those also incorrect?

File: CHANGELOG.md
Comment (pacu) [inline CHANGELOG.md:37]:
```suggestion
- `zcashlc_list_transparent_receivers`

```

File: rust/src/lib.rs
Comment (str4d) [reply rust/src/lib.rs:2149]: Probably yes. Can you give links?

File: rust/Cargo.toml
Comment (pacu) [inline rust/Cargo.toml:40]: I think these patches should be removed and point to creates only

**Review by pacu — CHANGES_REQUESTED** [review]

File: rust/Cargo.toml
Comment (sellout) [reply rust/Cargo.toml:40]: Yeah, @str4d mentioned this in our standup, so I’m already making that change.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:2149]: * https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/4485728196b5348759ceab2d23f3546d8708dc1a/rust/src/lib.rs#L1253-L1255
* https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/4485728196b5348759ceab2d23f3546d8708dc1a/rust/src/lib.rs#L1307-L1309
* https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/4485728196b5348759ceab2d23f3546d8708dc1a/rust/src/lib.rs#L1376-L1378
* https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/4485728196b5348759ceab2d23f3546d8708dc1a/rust/src/lib.rs#L1428-L1430

**Review by pacu — APPROVED** [review]

File: CHANGELOG.md
Comment (daira) [inline CHANGELOG.md:79]: "0 confirmations" normally means mempool transactions, but "up to those in the latest block" suggests otherwise (i.e. 1 confirmation or more). Please clarify.

File: rust/Cargo.lock
Comment (daira) [inline rust/Cargo.lock:128]: Filed https://github.com/zcash/librustzcash/issues/702 to consider eliminating the `base64` dependency in favour of `base64ct`.

File: rust/Cargo.lock
Comment (daira) [inline rust/Cargo.lock:1043]: Adding `orchard` as a dependency requires that a license exception notice be added to `zcash-light-client-ffi`'s `README.md`. See https://github.com/zcash/librustzcash/tree/main/zcash_client_backend for the wording. This is blocking.

**Review by daira — CHANGES_REQUESTED** [review]
License exception text needed in `README.md`.

File: rust/Cargo.lock
Comment (daira) [inline rust/Cargo.lock:1132]: Non-blocking: The duplicate dependency is [due to `schemer`](https://github.com/aschampion/schemer/blob/main/schemer/Cargo.toml#L19). If `schemer` updated to depend on `daggy 0.8` which depends on `petgraph 0.6`, this would be resolved.

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:242]: Nit:
```suggestion
/// have been received by the currently available account (in order to enable
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:263]:
```suggestion
pub unsafe extern "C" fn zcashlc_create_account(
```

File: CHANGELOG.md
Comment (sellout) [reply CHANGELOG.md:79]: The code is certainly doing `0` confirmations, and that does look like it includes mempool tx, as you say.

Since @str4d had commented on that code earlier and said “with a confirmations of `0`, the two should be identical,” I’m going to assume that means the mempool tx are intentional and update the changelog.

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:367]: Nit:
```suggestion
///   It must point to an array of `FFIEncodedKey` values.
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:489]:
```suggestion
pub unsafe extern "C" fn zcashlc_spending_key_to_full_viewing_key(
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:524]:
```suggestion
pub unsafe extern "C" fn zcashlc_init_blocks_table(
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:482]:
```suggestion
/// and returns the resulting encoded UFVK string. `usk_ptr` should point to an array of `usk_len`
/// bytes containing a unified spending key encoded as returned from the `zcashlc_create_account`
/// or `zcashlc_derive_spending_key` functions.
///
/// # Safety
///
/// - `usk_ptr` must be non-null and must point to an array of `usk_len` bytes.
```
The encoding is not a safety requirement, I think, only a correctness requirement.

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:451]:
```suggestion
/// across the FFI. `usk_ptr` should point to an array of `usk_len` bytes containing
/// a unified spending key encoded as returned from the `zcashlc_create_account` or
/// `zcashlc_derive_spending_key` functions. Callers should reproduce the following
/// safety documentation.
///
/// # Safety
///
/// - `usk_ptr` must be non-null and must point to an array of `usk_len` bytes.
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:520]:
```suggestion
/// The string represented by `sapling_tree_hex` should contain the encoded byte representation
/// of a Sapling commitment tree.
///
/// # Safety
///
/// - `db_data` must be non-null and valid for reads for `db_data_len` bytes, and it must have an
///   alignment of `1`. Its contents must be a string representing a valid system path in the
///   operating system's preferred representation.
/// - The memory referenced by `db_data` must not be mutated for the duration of the function call.
/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
///   documentation of pointer::offset.
/// - `hash_hex` must be non-null and must point to a null-terminated UTF-8 string.
/// - The memory referenced by `hash_hex` must not be mutated for the duration of the function call.
/// - `sapling_tree_hex` must be non-null and must point to a null-terminated UTF-8 string.
```

File: rust/Cargo.toml
Comment (daira) [inline rust/Cargo.toml:6]:
```suggestion
]
license = "MIT"
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:571]:
```suggestion
pub unsafe extern "C" fn zcashlc_get_current_address(
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:617]:
```suggestion
pub unsafe extern "C" fn zcashlc_get_next_available_address(
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:664]:
```suggestion
pub unsafe extern "C" fn zcashlc_list_transparent_receivers(
```

**Comment by daira** [comment]
This PR isn't compiling for me. Is `cd rust; cargo check` intended to work?

**Comment by sellout** [comment]
> This PR isn't compiling for me. Is `cd rust; cargo check` intended to work?

Yikes. It is certainly not compiling. I didn’t realize this repo doesn’t have any checks, and the build instructions in the README aren’t very helpful. E.g., `make && make install` doesn’t even build the package and `make xcframework` is macOS-specific. I’ll follow-up this PR with build & CI improvements. Looks like that’s already issue #2.

File: rust/Cargo.toml
Comment (sellout) [reply rust/Cargo.toml:40]: I’m going to do this in a separate PR.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:489]: Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:524]: Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:571]: Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:617]: Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:664]: Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:163]: Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:263]: Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs:372]: Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

File: rust/Cargo.lock
Comment (sellout) [reply rust/Cargo.lock:1130]: Made a PR: https://github.com/aschampion/schemer/pull/17

File: rust/Cargo.toml
Comment (sellout) [reply rust/Cargo.toml:42]: I’m working on this PR right now, _should_ be up for review tonight.

File: rust/src/lib.rs
Comment (daira) [reply rust/src/lib.rs:667]: Further `unsafe` annotations not commented because that will be addressed in a future PR.

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:786]:
```suggestion
/// - `ua` must be non-null and must point to a null-terminated UTF-8 string.
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:828]:
```suggestion
/// - `ua` must be non-null and must point to a null-terminated UTF-8 string.
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:907]:
```suggestion
    /// This instance produces no errors.
```
The empty `enum` is the idiomatic way of doing this in Rust, I think.

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:1095]:
```suggestion
/// - `ufvk` must be non-null and must point to a null-terminated UTF-8 string.
/// - The memory referenced by `ufvk` must not be mutated for the duration of the
///   function call.
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:1116]:
```suggestion
/// - `address` must be non-null and must point to a null-terminated UTF-8 string.
/// - The memory referenced by `address` must not be mutated for the duration of the
///   function call.
```

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:1528]: Document safety requirements on this helper function (similar to other helper functions above).

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:1896]: Incomplete sentence.

File: rust/src/lib.rs
Comment (daira) [inline rust/src/lib.rs:2070]:
```suggestion
/// - `s` should be a non-null pointer returned as a string by another zcashlc function.
```

**Review by daira — APPROVED** [review]
utACK modulo comments.

File: rust/src/lib.rs
Comment (sellout) [reply rust/src/lib.rs]: Looks like there’s also https://docs.rs/void/latest/void/, which has some useful related functions, like extracting the guaranteed value out of a `Result` that has `Void` on one side and `unreachable` (which is generally called `absurd` … but I like `unreachable`).

File: rust/src/lib.rs
Comment (nuttycom) [reply rust/src/lib.rs:163]: @str4d has recommended using the existing style, with unsafe blocks in the function body, so that it's easier to identify exactly which parts of the implementation involve unsafe calls. cc/ @daira @sellout

File: rust/Cargo.lock
Comment (sellout) [reply rust/Cargo.lock:1130]: Upstream PR is merged already, so just waiting on a release.

**Review by daira — COMMENTED** [review]
Post-hoc utACK 20966540d0cadf17a22457c9ef6260a137d002b6

File: rust/src/lib.rs
Comment (str4d) [reply rust/src/lib.rs:163]: Daira informed me that the lint referred to above enables that style to be maintained, but it was only added to stable releases relatively recently (which is why I didn't know about it). In applications, where we just follow latest stable, we should be able to use it.

**Pull request merged**
